src/snd/pulse/intercept/concrete.rs

```diff
@@ -0,0 +1,609 @@
+        if let Some(intercepted) = self.intercepted.as_ref() {
+            if source_idx != intercepted.index {
+                return Ok(false);
+            }
+
+            debug!("Stopping PulseAudio stream");
+            self.stream_manager.stop().await;
+            info!("Stopped monitor of application `{}'", intercepted);
+            self.intercepted = None;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn needs_monitor(&self) -> bool {
+        self.stream_manager.is_running()
+    }
+
+    async fn monitor(&mut self) {
+        self.stream_manager.monitor().await;
+    }
+
+    fn len(&self) -> usize {
+        if self.intercepted.is_some() { 1usize } else { 0usize }
+    }
+
+    fn stream_closed(&self) -> bool {
+        self.stream_manager.is_closed()
+    }
+
+    async fn close(mut self) {
+        self.stream_manager.stop().await;
+    }
+}
+
+impl LimitingInterceptor for SingleInputMonitor {
+    fn capacity(&self) -> Option<usize> {
+        Some(1usize)
+    }
+}
+
+#[async_trait]
+impl Interceptor for CapturingInterceptor {
+    async fn record<'a>(
+        &mut self,
+        source: Cow<'a, OwnedSinkInputInfo>
+    ) -> Result<(), InterceptError<'a>> {
+        if self.is_full() {
+            // limit is guaranteed to be Some if is_full() returns true
+            return Err(InterceptError::AtCapacity(
+                self.limit.unwrap(),
+                source
+            ));
+        }
+
+        if !self.stream_manager.is_running() {
+            debug!("Starting PulseAudio stream");
+            self.stream_manager.start(
+                self.rec.clone(),
+                self.volume
+            ).await?;
+        }
+
+        self.introspect.move_sink_input_by_index(
+            source.as_ref().index,
+            self.rec.index
+        ).await?;
+
+        self.intercepts.insert(source.index, InputState {
+            orig_sink: source.as_ref().sink,
+            corked: source.as_ref().corked,
+        });
+
+        info!("Intercepted application `{}'", source);
+
+        Ok(())
+    }
+
+    async fn update_intercept(
+        &mut self,
+        source: &OwnedSinkInputInfo
+    ) -> Result<(), Code> {
+        if let Some(intercept) = self.intercepts.get_mut(&source.index) {
+            if intercept.corked == source.corked {
+                return Ok(());
+            }
+
+            intercept.corked = source.corked;
+        } else {
+            return Ok(());
+        }
+
+        self.stream_manager.set_cork(
+            self.intercepts.values().any(|intercept| intercept.corked)
+        ).await
+    }
+
+    async fn stop(&mut self, source_idx: u32) -> Result<bool, Code> {
+        if let Some(intercept) = self.intercepts.get(&source_idx) {
+            let target_sink = util::get_sink_or_default(
+                &self.introspect,
+                intercept.orig_sink
+            ).await?;
+
+            self.introspect.move_sink_input_by_index(
```

Comment:
This probably needs to handle the case where `source_idx` refers to a sink input that no longer exists.